fw/t/unit/test_http_parser_common.h

```diff
@@ -672,48 +734,169 @@ do {										\
  */
 #define ASSIGN_FRAMES_FOR_H2(frames_definition)					\
 do {										\
-	INIT_FRAMES();								\
+	tfw_init_frames();							\
 	SET_FRAMES_BUF(frames_buf);						\
 	frames_definition;							\
 	RESET_FRAMES_BUF();							\
 } while(0)
 
 DECLARE_FRAMES_BUF(frames_buf, 3 * 1024);
 
+#ifdef CHUNK_MODE_DETACHED
+static tfw_test_chunk_t *tfw_prep_chunks(uint32_t chunk_cnt, uint32_t chunk_size,
+						uint32_t str_len)
+{
+	int i;
+	uint32_t last_chunk_len;
+	tfw_test_chunk_t *chunks;
+
+	TEST_DBG4("%s: [%u] chunks, chunk size %u, pg order %u, str len %u\n",
+		__func__, chunk_cnt, chunk_size, get_order(chunk_size), str_len);
+
+	/* temprorarily allow sleeping */
+	kernel_fpu_end();
+	chunks = (tfw_test_chunk_t *)__vmalloc(chunk_cnt * sizeof(*chunks),
+						__GFP_ZERO);
+	if (!chunks) {
+		TEST_DBG("%s: Failed to allocate chunk descriptors\n", __func__);
+		return ERR_PTR(-ENOMEM);
+	}
+	kernel_fpu_begin();
+	TEST_DBG4("%s: chunks buf %pK\n", __func__, chunks);
+
+	if (chunk_cnt > 1) {
+		for (i = 0; i < chunk_cnt - 1; i++) {
+			chunks[i].buf = (char *)kmalloc(chunk_size, GFP_ATOMIC);
+			if (!chunks[i].buf) {
+				TEST_DBG("%s: Failed to allocate chunk page(s)\n",
+					__func__);
+				goto err_page_alloc;
+			}
+		}
+	}
+
+	/* last chunk size may differ, so set it up separately */
+	last_chunk_len = str_len % chunk_size;
+	if (last_chunk_len != 0) {
+		chunk_size = last_chunk_len;
+	}
+
+	chunks[chunk_cnt - 1].buf = (char *)kmalloc(chunk_size, GFP_ATOMIC);
+	if (!chunks[chunk_cnt - 1].buf) {
+		TEST_DBG("%s: Failed to allocate last chunk page(s)\n", __func__);
+		goto err_page_alloc;
+	}
```

Comment:
I think we might get the same problem as before and, worse, get the tests blinking depending on how `kmalloc()` allocates memory - if (e.g. on a fresh system) it allocates all the chunks from sibling buddies, then we're in exactly the same case as before.

I propose to do something like
```C
char *tmp_buf = (char *)kmalloc(chunk_size + 16, GFP_ATOMIC);
memset(tmp_buf, 0x55, chunk_size + 16);
chunks[...].buf = tmp_buf + 8;
```

Maybe we also should consider to set non-aligned addresses for chunks - TBD.

Reply:
OK, I see the requirement for KASAN, so with KASAN this poisoning might be not necessary. However, IMO the tests should be correct and efficient regardless kernel options. Otherwise with each parser change, we'll need a KASAN-enabled kernel.

Reply:
I believe using KASAN infra is more reliable than just adding guard buffers to the payload.
If I understood correctly, allocating chunks from sibling buddies is not an issue to KASAN. It would still be able to track out-of-bounds accesses and so on. More than that, KASAN can produce pretty informative stack traces to find where the problem is.
We could technically always run the CI tests on a VM with KASAN enabled kernel. It would definitely slow everything down a little, but it's still worth it IMHO.